lib/Scheduling/TestPasses.cpp

```diff
@@ -53,26 +63,39 @@ static LogicalResult constructProblem(Problem &prob, FuncOp func) {
     }
   }
 
-  // parse auxiliary dependences in the testcase, encoded as an array of
-  // 2-element arrays of integer attributes (see `test_asap.mlir`)
+  // parse auxiliary dependences in the testcase, encoded as an array of arrays
+  // of integer attributes
   if (auto attr = func->getAttrOfType<ArrayAttr>("auxdeps")) {
     auto &ops = prob.getOperations();
-    for (auto auxDepAttr : attr.getAsRange<ArrayAttr>()) {
-      if (auxDepAttr.size() != 2)
-        continue;
-      auto fromIdxAttr = auxDepAttr[0].dyn_cast<IntegerAttr>();
-      auto toIdxAttr = auxDepAttr[1].dyn_cast<IntegerAttr>();
-      if (!fromIdxAttr || !toIdxAttr)
-        continue;
-      unsigned fromIdx = fromIdxAttr.getInt();
-      unsigned toIdx = toIdxAttr.getInt();
-      if (fromIdx >= ops.size() || toIdx >= ops.size())
-        continue;
-
-      // finally, we have two integer indices in range of the operations list
-      if (failed(prob.insertDependence(
-              std::make_pair(ops[fromIdx], ops[toIdx])))) {
+    unsigned nOps = ops.size();
+    for (auto &elemArr : parseArrayOfArrays(attr)) {
+      if (elemArr.size() < 2 || elemArr[0] >= nOps || elemArr[1] >= nOps)
+        continue; // malformed testcase
+
+      Operation *from = ops[elemArr[0]];
+      Operation *to = ops[elemArr[1]];
+      if (failed(prob.insertDependence(std::make_pair(from, to))))
         return func->emitError("inserting aux dependence failed");
+    }
+  }
+
+  return success();
+}
+
+static LogicalResult constructCyclicProblem(CyclicProblem &prob, FuncOp func) {
+  if (failed(constructProblem(prob, func)))
+    return failure();
+
+  // parse auxiliary dependences in the testcase (again), in order to set the
+  // optional distance in the cyclic problem
+  if (auto attr = func->getAttrOfType<ArrayAttr>("auxdeps")) {
+    auto &ops = prob.getOperations();
+    for (auto &elemArr : parseArrayOfArrays(attr)) {
+      if (elemArr.size() == 3) {
```

Comment:
Could this also just assert `elemArr.size() == 3`?

Reply:
I want the distance to be optional in the test cases, too, so here I really need to just skip instead of `assert`'ing. (I did change the code to `continue` early when the distance is not specified, though.)